parlai/scripts/train_model.py

```diff
@@ -242,6 +241,45 @@ def load_eval_worlds(agent, opt, datatype):
     return worlds
 
 
+def _run_single_eval(opt, valid_world, max_exs):
+    # run evaluation on a single world
+    valid_world.reset()
+
+    cnt = 0
+    max_cnt = max_exs if max_exs > 0 else float('inf')
+    while not valid_world.epoch_done() and cnt < max_cnt:
+        valid_world.parley()
+        if cnt == 0 and opt['display_examples']:
+            print(valid_world.display() + '\n~~')
+            print(valid_world.report())
+        cnt += valid_world.opt['batchsize']
+
+    valid_report = valid_world.report()
+    valid_world.reset()  # make sure world doesn't remember valid data
+
+    return valid_report
+
+
+def _aggregate_reports(reports, valid_worlds):
+    if len(reports) == 1:
+        return reports[0]
+
+    # multiple tasks, aggregate metrics
+    tasks = [world.opt['task'] for world in valid_worlds]
+    agent_metrics = {}
+    total_report = {'tasks': {}}
+    for i, report in enumerate(reports):
+        total_report['tasks'][tasks[i]] = report
+        for metric, val in report.items():
+            agent_metrics.setdefault(metric, []).append(val)
+    total_report['tasks']['all'] = {}
+    for metric, vals in agent_metrics.items():
+        total_report['tasks']['all'][metric] = round_sigfigs(
+            sum(vals) / len(vals), 4)
```

Comment:
so much lost precision from round_sigfigs at every level (no change needed in this patch but SAD!)